### ISSUE TITLE
InvalidOperationException in GetNotEmptyFormatterExtensionNames()

### ISSUE THREAD

**Issue by drmcclelland** [scottrippey/SmartFormat.NET#103]
My code succesufully uses SmartFormatter Format() many hundreds of times per day in an ASP.NET WebAPI webservice - but occasionally (every 2 weeks or so) I log the following exception which I have been never been able to reproduce in my development environment.  

Does anyone have an idea what may be going on?

```System.InvalidOperationException: Collection was modified; enumeration operation may not execute.     at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)     at System.Collections.Generic.List`1.Enumerator.MoveNextRare()     at SmartFormat.SmartFormatter.GetNotEmptyFormatterExtensionNames()     at SmartFormat.SmartFormatter.Format(IFormatProvider provider, String format, Object[] args)     at SmartFormat.Smart.Format(String format, Object arg0)```

**Comment by axunonb** [comment]
Just seeing the exception message this is hard to say, and I've never seen this exception before.
It looks you're using the ```ListFormatter``` extension. When working with nested lists or arrays, we're keeping track of the collection index. Could have sth. to do with this in a multi-threaded environment. Does this help to reproduce the exception?

**Comment by drmcclelland** [comment]
Thanks for the hints - I am using this in an ASP.NET WebAPI project.  I'm adding my own extensions, but I think I'm doing it incorrectly for a multi-threaded environment:

```
Smart.Default.AddExtensions(new DollarsToCentsFormatter(), new DictionaryLookupFixedWidthFormatter(), new RandomNumberFormatter());

foreach (var kvp in fieldsToCheck)
{
    message[kvp.Key] = Smart.Format(kvp.Value, this);
}

```

I can see how if another request comes in that it could conflict since Smart.Default is a static instance variable.

Instead, I'm going to try this:

```
var smartFormatter = Smart.CreateDefaultSmartFormat();
smartFormatter.AddExtensions(new DollarsToCentsFormatter(), new DictionaryLookupFixedWidthFormatter(), new RandomNumberFormatter());

foreach (var kvp in fieldsToCheck)
{
    message[kvp.Key] = smartFormatter.Format(kvp.Value, this);
}

```

Whaddya think?

**Comment by axunonb** [comment]
Yep, creating a new instance of ```SmartFormat``` is the right choice in a multi-threaded environment. Have in mind to include all extensions you need,
